buckets/local/bucket_test.go

```diff
@@ -1,302 +1,516 @@
+	err = os.RemoveAll(fpth)
+	_, err = buck.PushLocal(context.Background())
 	require.Nil(t, err)
-	assert.True(t, lc.Defined())
 
-	err = buck.SetRemotePath("", lc)
+	var buf bytes.Buffer
+	err = buck.DecryptRemotePath(context.Background(), "ciphertext", "badpass", &buf)
+	require.NotNil(t, err)
+
+	err = buck.DecryptRemotePath(context.Background(), "ciphertext", "shhhhh!", &buf)
 	require.Nil(t, err)
+
+	assert.Equal(t, 1024, buf.Len())
 }
 
-func TestBucket_MatchPath(t *testing.T) {
-	buck := makeBucket(t, "testdata/a", options.BalancedLayout)
-	defer buck.Close()
+func TestBucket_AddRemoteCid(t *testing.T) {
+	t.Parallel()
+	buckets := setup(t)
+	conf := getConf(t, buckets)
+	buck, err := buckets.NewBucket(context.Background(), conf)
+	require.Nil(t, err)
+
+	pth := createIpfsFolder(t)
+	err = buck.AddRemoteCid(context.Background(), pth.Cid(), conf.Path)
+	require.Nil(t, err)
+
+	bp, err := buck.Path()
+	require.Nil(t, err)
+	_, err = os.Stat(filepath.Join(bp, "file1.txt"))
+	require.Nil(t, err)
+	_, err = os.Stat(filepath.Join(bp, "folder1", "file2.txt"))
+	require.Nil(t, err)
+}
+
+func TestBucket_DiffLocal(t *testing.T) {
+	t.Parallel()
+	buckets := setup(t)
+	conf := getConf(t, buckets)
+	buck, err := buckets.NewBucket(context.Background(), conf)
+	require.Nil(t, err)
+
+	addRandomFile(t, buck, "file1", 256)
+	fpth := addRandomFile(t, buck, "folder/file2", 256)
+	diff, err := buck.DiffLocal()
+	require.Nil(t, err)
+	assert.Len(t, diff, 2)
+	assert.Equal(t, diff[0].Type, dagutils.Add)
+
+	// Test diff from a nested folder
+	buck2, err := buckets.GetLocalBucket(context.Background(), filepath.Join(conf.Path, "folder"))
+	require.Nil(t, err)
+	diff, err = buck2.DiffLocal()
+	require.Nil(t, err)
+	assert.Len(t, diff, 2)
+	assert.Equal(t, diff[0].Type, dagutils.Add)
+
+	_, err = buck.PushLocal(context.Background())
+	require.Nil(t, err)
+
+	diff, err = buck.DiffLocal()
+	require.Nil(t, err)
+	assert.Len(t, diff, 0)
+
+	addRandomFile(t, buck, "file1", 256)
+	diff, err = buck.DiffLocal()
+	require.Nil(t, err)
+	assert.Len(t, diff, 1)
+	assert.Equal(t, diff[0].Type, dagutils.Mod)
+
+	_, err = buck.PushLocal(context.Background())
+	require.Nil(t, err)
+
+	diff, err = buck.DiffLocal()
+	require.Nil(t, err)
+	assert.Len(t, diff, 0)
 
-	err := buck.Save(context.Background())
+	err = os.RemoveAll(fpth)
 	require.Nil(t, err)
-	lc, _, err := buck.Root()
+	diff, err = buck.DiffLocal()
 	require.Nil(t, err)
-	assert.True(t, lc.Defined())
+	assert.Len(t, diff, 1)
+	assert.Equal(t, diff[0].Type, dagutils.Remove)
 
-	rc := makeCid(t, "remote")
-	err = buck.SetRemotePath("", rc)
+	_, err = buck.PushLocal(context.Background())
 	require.Nil(t, err)
 
-	match, err := buck.MatchPath("", lc, rc)
+	diff, err = buck.DiffLocal()
 	require.Nil(t, err)
-	assert.True(t, match)
+	assert.Len(t, diff, 0)
 }
 
-func TestBucket_RemovePath(t *testing.T) {
-	buck := makeBucket(t, "testdata/a", options.BalancedLayout)
-	defer buck.Close()
+func TestBucket_Watch(t *testing.T) {
```

Comment:
Here's an interesting test.
- Create a bucket, watch it, and add a file to it's dir
- It should auto-push
- Join the same bucket from a different thread host
- It should have the file
- Add a file to the second bucket and push it
- The first bucket should have pulled it down from the thread change event